### ISSUE TITLE
XLA jit_compile=True returns different argmax index on NaN input than eager execution

### ISSUE THREAD

**Issue by ALinrunrun** [tensorflow/tensorflow#118177]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.21.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Python version

3.11

### Current behavior?

TensorFlow eager and XLA return different results for tf.math.argmax when the input contains NaN.

The input tensor is [4.0, 5.5, NaN]. Eager execution returns index 1, corresponding to the finite value 5.5. The same function compiled with jit_compile=True returns index 2, corresponding to the NaN value.

In the reproducer below, eager returns 1, while XLA returns 2.

Expected behavior?

tf.math.argmax should handle NaN consistently between eager execution and jit_compile=True for the same input.

### Standalone code to reproduce the issue

```shell
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "-1"

import tensorflow as tf
import numpy as np


t = tf.constant([4.0, 5.5, np.nan], dtype=tf.float32)

eager = int(tf.math.argmax(t).numpy())
jit = int(tf.function(lambda v: tf.math.argmax(v), jit_compile=True)(t).numpy())

print(f"eager argmax: {eager}")
print(f"jit   argmax: {jit}")

if eager != jit:
    print("BUG REPRODUCED: argmax NaN behavior differs between eager and jit_compile=True")
else:
    print("not reproduced")
```

### Relevant log output

```shell
eager argmax: 1
jit argmax: 2
BUG REPRODUCED: argmax NaN behavior differs between eager and jit_compile=True
```

**Comment by avmohansai** [comment]
Hi [@Venkat6871](https://github.com/Venkat6871) ,
I Was able to reproduce the issue on Colab using `TF v2.21.0` and `TF-nightly` ,Please find the [gist](https://colab.sandbox.google.com/gist/avmohansai/4e509409a226c2335c34e6de36259052/118177_tf_2-21-0_tf-nightly.ipynb) here for reference.
Thank you.

**Comment by dmiltr3** [comment]
Hi @ALinrunrun, thanks for reporting this and providing the standalone reproduction script.

The discrepancy you observed on `t = tf.constant([4.0, 5.5, np.nan])` between Eager (returning `1`) and XLA `jit_compile=True` (returning `2`) is understandable to flag. However, after investigating the underlying implementations, **XLA's behavior (`2`) is intentional, follows standard numerical computing conventions, and is working as designed**. 

We recommend closing this issue for the following reasons:

---

### 1. XLA (`jit_compile=True`) Matches NumPy, JAX, and PyTorch

Across standard Python numerical computing libraries, `argmax` on your exact input returns index `2` (the index of `NaN`), rather than silently dropping it:

```python
import numpy as np
import torch
import jax.numpy as jnp

# NumPy:
np.argmax([4.0, 5.5, np.nan])
# Output: 2

# JAX:
jnp.argmax(jnp.array([4.0, 5.5, np.nan]))
# Output: Array(2, dtype=int32)

# PyTorch:
torch.argmax(torch.tensor([4.0, 5.5, float('nan')]))
# Output: tensor(2)
```

In numerical pipelines and ML model training, a `NaN` indicates numerical instability (e.g., gradient explosion, invalid division, or overflow). Standard `argmax` intentionally surfaces the `NaN` rather than silently discarding it and returning an arbitrary finite value, which could mask serious corruption in model state.

When users specifically want to ignore `NaN` values, dedicated functions are used instead (such as NumPy's `np.nanargmax([4.0, 5.5, np.nan])`, which returns `1`).

---

### 2. Eager Mode Returning `1` is an Accidental Artifact of IEEE 754 Comparisons

It might seem that Eager mode deliberately implements a "skip NaN" policy, but this is actually an accidental consequence of Eigen's C++ reduction kernel (`third_party/tensorflow/core/kernels/argmax_op.h`).

Eigen computes `argmax` by scanning the elements with standard C++ `>` comparisons:
```cpp
if (current_value > max_value) {
  max_value = current_value;
  max_index = current_index;
}
```

Under IEEE 754 floating-point rules, **any comparison involving `NaN` evaluates to `false`** (`x > NaN` is `false`, and `NaN > x` is `false`).

You can see this by slightly modifying the order of elements in your reproducer:

```python
import tensorflow as tf
import numpy as np

# Your original input: NaN at the end
t1 = tf.constant([4.0, 5.5, np.nan])
print("Eager [4.0, 5.5, NaN]:", int(tf.math.argmax(t1).numpy()))
# Output: 1  (5.5 > 4.0 is True; NaN > 5.5 is False -> keeps 5.5)

# Same values, but NaN at the beginning:
t2 = tf.constant([np.nan, 4.0, 5.5])
print("Eager [NaN, 4.0, 5.5]:", int(tf.math.argmax(t2).numpy()))
# Output: 0  (accumulator starts at NaN; 4.0 > NaN is False; 5.5 > NaN is False -> returns 0!)
```

Notice that for `[NaN, 4.0, 5.5]`, Eager execution returns index `0` (the `NaN`). Eager does not actually ignore `NaN`s—its output depends arbitrarily on whether the `NaN` appears before or after the maximum due to unordered comparisons in `>`.

---

### 3. Why XLA Should Not Be Changed

XLA is the shared compiler backend for TensorFlow, JAX, and OpenXLA. Its reduction comparator is specifically implemented to provide consistent, deterministic handling that does not silently swallow `NaN` values. Changing XLA's `ArgMinMax` reduction to drop NaNs would:
1. Break consistency with NumPy, JAX (`lax.argmax`), and PyTorch.
2. Silently mask numerical errors during accelerated model execution.

---

### 4. Recommended Workaround

If your application specifically requires ignoring `NaN` values (i.e. `nanargmax` semantics), the recommended approach is to explicitly mask `NaN` values before calling `tf.math.argmax`. This produces identical, deterministic results across both Eager and `jit_compile=True`:

```python
import tensorflow as tf
import numpy as np

t = tf.constant([4.0, 5.5, np.nan], dtype=tf.float32)

def nanargmax(v, axis=-1):
  # Replace NaNs with -inf so they are never selected by argmax
  sanitized = tf.where(tf.math.is_nan(v), tf.constant(-float("inf"), dtype=v.dtype), v)
  return tf.math.argmax(sanitized, axis=axis)

eager_res = int(nanargmax(t).numpy())
jit_res   = int(tf.function(nanargmax, jit_compile=True)(t).numpy())

print(f"eager nanargmax: {eager_res}")  # 1
print(f"jit   nanargmax: {jit_res}")    # 1
```

Because XLA's behavior is consistent with the broader ecosystem and avoids silent failure modes, we believe XLA is working as intended here and propose closing this issue.
